src/main/java/tempest/GUI/ErrorMessage.java

```diff
@@ -5,11 +5,11 @@
 public class ErrorMessage{
     private static boolean messagesShown = true;
 
-    public ErrorMessage(){}
-
     public void showMessage(JPanel parent, String message){
```

Comment:
Might it be better to pass an error object in here to allow for easier passing of generic errors?

Reply:
An Error object or an Exception object?

Either way, since we're showing these messages to the user, wouldn't having the pop up contain a number format exception be a bit confusing - it wouldn't say for example "Invalid minutes entered"

Reply:
Sure, but you can extract that from the message inside the exception.

Reply:
I set an Exception as the parameter in showMessage(), so when calling showMessage you create a new exception object e.g. new NumberFormateException() - The message you get from this is null, so you have to enter a message in the brackets e.g. new NumberFormateException("Invalid Minutes Entered"). Which is just the same as we have currently, just without having to create an exception.

Reply:
That's right, but if we need to display default exceptions to the user elsewhere in the application, this interface makes the component more reusable.

Reply:
But to get the default message from an exception you have to catch it then call e.getMessage() - so can't you just call showMessage() with this string? 

That would be a lot simpler than changing showMessage to except exceptions (and having all the specific (not caught) exceptions be  changed to -> new Exception("example message") ). Also, inside showMessage() we'll be calling getMessage() anyways so calling getMessage() beforehand just for the defaults will mean we don't also have to call getMessage for our specific written messages.

Reply:
If we use exceptions more, the exception handling code can be centralised and avoids the mixing of approaches seen in `AddSessionPage` or `AddModulePage` whilst setting us up better for the future.

Reply:
What do you think we'll have more of by the end for the popups? Specific messages or caught exceptions where we use the default message?

If caught exceptions then sure I'll refactor the code

Reply:
"the mixing of approaches seen in AddSessionPage or AddModulePage" - What do you mean by this?

Reply:
# `AddSessionPage`
```Java
if (hours.equals("") && minutes.equals("")) { // Empty hours and minutes
            errorMessage("A session needs to be >= 1 minute");
            return;
        } else if (hours.equals("")) { // Only minutes have been entered
            try {
                minutesInt = Integer.parseInt(minutes);

                if (minutesInt <= 0) {
                    throw new NumberFormatException(); // Negative minutes
                }
            } catch (NumberFormatException e) {
                errorMessage("Invalid minutes entered");
                return;
            }
        } else if (minutes.equals("")) { // Only hours have been entered
            try {
                hoursInt = Integer.parseInt(hours);

                if (hoursInt <= 0) {
                    throw new NumberFormatException(); // Negative hours
                }
            } catch (NumberFormatException e) {
                errorMessage("Invalid hours entered");
                return;
            }
        } else { // Hours and minutes have been entered
            try {
                minutesInt = Integer.parseInt(minutes);
                hoursInt = Integer.parseInt(hours);

                if (hoursInt <= 0 || minutesInt <= 0) {
                    throw new NumberFormatException(); // At least one is negative
                }

                if (minutesInt > 59) {
                    errorMessage("If an hour is entered minutes should be < 60"); // Minutes exceeds 59 with an hour
                                                                                  // inputted
                    return;
                }
            } catch (NumberFormatException e) {
                errorMessage("Invalid time entered");
                return;
            }
        }
```
This bit uses a whole bunch of try catches that could be aggregated into one. Later on...
```Java
 // Checks the session entered is under 24 hours
        if (time.toMinutes() > 24 * 60) {
            errorMessage("A study session can't be longer than 24 hours");
            return;
        }

        Module module = null;
        boolean foundName = false;

        for (Module m : state.getModules()) {
            if (moduleName.equals(m.getName())) {
                module = m;
                foundName = true;
                break;
            }
        }

        if (!foundName) {
            errorMessage("Unable to find this module");
            return;
        }
```
This doesn't bother with try-catch, it just goes straight to error messages. This could all be simplified, reducing the calls to `errorMessage`, improving readability and increasing performance.

Reply:
But having all those separate try catches allows for better feedback to the user to explain what went wrong.

With the second code snippet, are you saying it should use a try catch? Or if not - how would this code be simplified?

Are you saying to basically have this entire method in a try catch (Exception e) - then (atm at least) sending e.getMessage() to errorMessage()?

Reply:
Yes, I think that would work.

Reply:
So I had a go at putting everything in a try-catch. I was able to simplify some stuff, meaning it ended up being 20 lines shorter. 

However, this means some input errors don't have a specifically written message. So for example entering letters causes the following:

![image](https://user-images.githubusercontent.com/65258846/112218574-ca3e0580-8c1b-11eb-929b-15a870dbb879.png)

So although this error is caught, it's not very helpful to the user.

I can print out the type of exception, but this would mean exceptions like entering negative numbers, which do have a specific message, would also have their exception type printed.

![image](https://user-images.githubusercontent.com/65258846/112218986-54866980-8c1c-11eb-8503-659c6214b67b.png)

![image](https://user-images.githubusercontent.com/65258846/112219030-623bef00-8c1c-11eb-97e8-9ea4d08fe562.png)

To get this all working I obviously changed `errorMessage()` and `showMessage()` to take Exceptions rather than Strings. But I still can't envisage when we would want to have a popup with a default exception message. It's just not helpful to the user.

And just because we're using pop up messages, doesn't mean we have to use them for everything, we can still print things to the console.

I think it's also worth remembering the spec says:

> your programming ability and code structure will not be assessed as part of this assignment

And so for a class like `AddSessionPage ` where it's finished and works, and shouldn't need to be changed in the future, it feels like there are better uses of our time .